File: src/BaekJoon/C++/baekjoon_10845.cpp
// https://www.acmicpc.net/problem/10845

#include <iostream>
#include <string>
using namespace std;

class queue {
private :
	class queueNode {
		friend class queue;
		int element;
		queueNode *next;
		queueNode(int element, queueNode *next) {
			this->element = element;
			(*this).next = next;
		}
	};
	queueNode *frontQueue;
	queueNode *rearQueue;
	int queueSize;

public :
	queue() {
		frontQueue = NULL;
		rearQueue = NULL;
		queueSize = 0;
	}
	void push(int x);
	int pop();
	int size();
	int empty();
	int front();
	int back();
};

void queue::push(int x) {
	queueSize++;
	queueNode *pushNode = new queueNode(x, NULL);
	if(frontQueue == NULL)
		frontQueue = pushNode;
	else
		rearQueue->next = pushNode;
	rearQueue = pushNode;
}
int queue::pop() {
	if(frontQueue == NULL)
		return -1;
	else {
		queueSize--;
		int temp = frontQueue->element;
		frontQueue = frontQueue->next;
		if(empty())
			rearQueue = NULL;
		return temp;
	}
}
int queue::size() {
	return queueSize;
}
int queue::empty() {
	if(queueSize == 0)
		return 1;
	else
		return 0;
}
int queue::front() {
	if(frontQueue == NULL)
		return -1;
	else
		return frontQueue->element;
}
int queue::back() {
	if(rearQueue == NULL)
		return -1;
	else
		return rearQueue->element;
}

int main() {
	queue q;
	int n;

	cin >> n;

	for(int i = 0; i < n; i++) {
		string str;
		cin >> str;

		if(str == "push") {
			int x;
			cin >> x;
			q.push(x);
		} else if(str == "pop") {
			cout << q.pop() << "\n";
		} else if(str == "size") {
			cout << q.size() << "\n";
		} else if(str == "empty") {
			cout << q.empty() << "\n";
		} else if(str == "front") {
			cout << q.front() << "\n";
		} else if(str == "back") {
			cout << q.back() << "\n";
		}

	} // end of for
	return 0;
}

File: src/BaekJoon/java/baekjoon_10845.java
package BaekJoon.java;
// https://www.acmicpc.net/problem/10845

import java.math.BigDecimal;
import java.util.Scanner;
import java.util.StringTokenizer;

class Queue_Node {
	Queue_Node next;
	int data;

	Queue_Node() {
	}

	Queue_Node(int element) {
		this.data = element;
	}

	Queue_Node(int element, Queue_Node next) {
		this.data = element;
		this.next = next;
	}
}

class Queue {
	Queue_Node front;
	Queue_Node rear;
	int size = 0;

	Queue() {
		front = rear = null;
	}

	void push(int x){
		size++;
		Queue_Node p = new Queue_Node(x, null);

		if(front == null)
			front = p;
		else
			rear.next = p;
		rear = p;
	}

	int pop() {
		if(front == null)
			return -1;
		else {
			size--;
			int data = front.data;
			front = front.next;
			if (empty() == 1)
				rear = null;
			return data;
		}
	}

	int empty() {
		if(front == null)
			return 1;
		else
			return 0;
	}

	int front() {
		if(front == null)
			return -1;
		return front.data;
	}

	int rear() {
		if(rear == null)
			return -1;
		return rear.data;
	}
}

public class baekjoon_10845 {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		Queue queue = new Queue();

		for(int test_case = 1; test_case <= n; test_case++) {
			String s = sc.next();

			if(s.equals("push")) {
				int x = sc.nextInt();
				queue.push(x);
			} else if(s.equals("pop"))
				System.out.println(queue.pop());
			else if(s.equals("size"))
				System.out.println(queue.size);
			else if(s.equals("empty"))
				System.out.println(queue.empty());
			else if(s.equals("front"))
				System.out.println(queue.front());
			else if(s.equals("back"))
				System.out.println(queue.rear());
		}

		sc.close();
	}
}